feat(event): Adding eventing lib

robzienert opened the pull request:
This is PR 1 of N for some new idempotency functionality in River. The implementation will be confined to the Titus provider only while we iterate ergonomics of the various components.

The idempotency functionality implements a Saga pattern that is backed by an immutable event log. While not strictly required for the design pattern, backing operations by an eventing system will make it more palatable for debugging and auditing functionality later on, as well as new interaction possibilities between clouddriver & orca in the future.

Current implementation targets an in-memory store only. Future contributions will add a SQL backend so that operations can be retried by Orca in the face of downstream cloud provider failures.

It is possible that this functionality could be lifted into kork at some point.

robzienert commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/config/MemoryEventRepositoryConfig.kt:
```diff
@@ -0,0 +1,100 @@
+      return true
+    }
+  }
+}
```
Here's me, making an attempt at some better patterns for handling configuration within Spinnaker.

ajordens commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt:
```diff
@@ -0,0 +1,168 @@
+    if (aggregate.version != originatingVersion) {
+      // If this is being thrown, ensure that the originating process is retried on the latest aggregate version
+      // by re-reading the events list.
+      throw AggregateChangeRejectedException(
```
The expected retry behavior for this is something upstream correct and up to the caller right?

robzienert replied:
That's correct. Upstream could be code interacting with the repository, or bubbled all the way out of the service boundary.

ajordens commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt:
```diff
@@ -0,0 +1,168 @@
+    this.events[aggregate]!!.addAll(events)
+    aggregate.version = aggregate.version + 1
+
+    events.joinToString(",") { it.javaClass.simpleName }
```
Looks unnecessary, it's already in the log below explicitly.

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/SpinEvent.kt:
```diff
@@ -0,0 +1,41 @@
+  include = JsonTypeInfo.As.PROPERTY,
+  property = "spinEventType"
+)
+abstract class SpinEvent(
```
I prefer `SpinnakerEvent`

robzienert commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/config/MemoryEventRepositoryConfig.kt:
```diff
@@ -0,0 +1,100 @@
+import kotlin.reflect.KClass
+
+@Configuration
+@ConditionalOnProperty("spinnaker.clouddriver.eventing.memory-repository.enabled", matchIfMissing = true)
```
Ah, yes! Thanks for calling that out. I've been dabbling with trying to come up with a standard config pattern. Extracted this gist from some private gist I cooked up a few months ago while I was out of commission for some color: https://gist.github.com/robzienert/68dde8c895e93c0a391953eae43fcf0f

I was planning on submitting an RFC for this but never got around to it. Happy to debate this, I'm dropping the gist here mostly to get wheels turning.

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt:
```diff
@@ -0,0 +1,168 @@
+        "(version: ${aggregate.version}, originatingVersion: $originatingVersion)")
+    }
+
+    val currentSequence = this.events[aggregate]!!.map { it.metadata.sequence }.max() ?: 0
```
unless i am missing something, i don't think the sequence calculation is thread-safe

marchello2000 replied:
actually also the sequence assignment on L69 is not safe either unless this whole method is in a critical section (or equivalent)

robzienert replied:
My assumption was that an aggregate would be written to by a single thread. Probably a poor assumption, so I'll update.

marchello2000 replied:
ah, yeah, that makes sense - didn't realize it was meant that only 1 thread per aggregate

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt:
```diff
@@ -0,0 +1,168 @@
+
+  private val events: MutableMap<Aggregate, MutableList<SpinEvent>> = ConcurrentHashMap()
+
+  override fun save(aggregateType: String, aggregateId: String, originatingVersion: Long, events: List<SpinEvent>) {
```
nit: having `events` and `MemoryEventRepository.events` makes it harder to grok which is being operated on while reading this function. Maybe `newEvents` for the local var?

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/MemoryEventRepository.kt:
```diff
@@ -0,0 +1,168 @@
+    log.debug("Saved $aggregateType/$aggregateId@${aggregate.version}: " +
+      "[${events.joinToString(",") { it.javaClass.simpleName }}]")
+
+    events.forEach { eventPublisher.publish(it) }
```
given that the `eventPublisher` can be (as is now) synchronous, I think it's a slightly weird/unexpected for the caller of this method to have to wait on all publishers having to process the events being saved. I think the caller would expect the events to be saved durably but not necessarily broadcast on the same blocking operation

robzienert replied:
This is a result of the current Saga framework design. You raise a good point, so I'm working on refactoring the Saga lib to separate the flow propagation from being event-driven to just informed by event data. This will let me remove this specific call and allow event publishing to be handled entirely asynchronously. This will also have the side effect of making `SynchronousEventPublisher` basically useless (and therefore, I'll re-evaluate the default `EventPublisher` implementation).

Thanks!

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/SynchronousEventPublisher.kt:
```diff
@@ -0,0 +1,47 @@
+        log.trace("Publishing event: $event to ${it.joinToString(",")}")
+      }.forEach {
+        try {
+          it.onEvent(event)
```
Always scares me when things like this are synchronous - one bad actor/EventListener and everyone gets hurt

robzienert replied:
That's what the try/catch is for - it negates the exact scenario you're talking about.

marchello2000 replied:
what about one actor taking 10 minutes to process the event?

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/EventMetadata.kt:
```diff
@@ -0,0 +1,35 @@
+  val originatingVersion: Long,
+  val timestamp: Instant = Instant.now(),
+  val serviceVersion: String = "unknown",
+  val provenance: String = "unknown"
```
this might be just me, but i hate the word `provenance`. it always takes me a few seconds to think about what it means. I am going to go on a limb and say that, esp for developers whose primary language is not English, will find this work similarly difficult to parse.
I think `source` or `origin` convey the same information and are much more widely used

robzienert commented:
Cleaned a bunch of stuff up, including getting rid of the `EventPublisher` in favor of the Spring `ApplicationEventPublisher`, which is async. The Saga lib I've refactored to not be driven off synchronous retrieval of events - how it should've been the whole time. 😬

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/Aggregate.kt:
```diff
@@ -0,0 +1,49 @@
+
+  override fun hashCode(): Int {
+    var result = type.hashCode()
+    result = 31 * result + id.hashCode()
```
do you need mod `Int.MAX_VALUE`?

robzienert replied:
This is autogenerated.

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/SpinnakerEvent.kt:
```diff
@@ -0,0 +1,41 @@
+@JsonTypeInfo(
+  use = JsonTypeInfo.Id.NAME,
+  include = JsonTypeInfo.As.PROPERTY,
+  property = "spinEventType"
```
where is this prop defined?

robzienert replied:
In the JSON payload.

marchello2000 commented on clouddriver-event/src/main/kotlin/com/netflix/spinnaker/clouddriver/event/persistence/InMemoryEventRepository.kt:
```diff
@@ -0,0 +1,165 @@
+      }
+  }
+
+  @Scheduled(fixedRate = 1_000)
```
every second seems a bit aggressive, no? not sure what the actual usage is (how many events are coming /going).
do you think it's worth making this recurrence interval configurable?

marchello2000 commented:
LGTM
